java-frontend/src/main/java/org/sonar/java/resolve/LeastUpperBound.java

```diff
@@ -0,0 +1,336 @@
+    return typeSubstitutionSolver.applySubstitution((JavaType) type, substitution);
+  }
+
+  private static List<Set<Type>> erased(Iterable<Set<Type>> typeSets) {
+    List<Set<Type>> results = new ArrayList<>();
+    for (Set<Type> typeSet : typeSets) {
+      Set<Type> erasedTypes = new LinkedHashSet<>();
+      for (Type type : typeSet) {
+        erasedTypes.add(type.erasure());
+      }
+      results.add(erasedTypes);
+    }
+    return results;
+  }
+
+  private static List<Type> intersection(List<Set<Type>> supertypes) {
+    List<Type> results = new ArrayList<>(supertypes.get(0));
+    for (int i = 1; i < supertypes.size(); i++) {
+      results.retainAll(supertypes.get(i));
+    }
+    return results;
+  }
+
+  /**
+   * Let MEC, the minimal erased candidate set for U1 ... Uk, be:
+   * MEC = { V | V in EC, and for all W != V in EC, it is not the case that W <: V }
+   * @param erasedCandidates
+   * @return
+   */
+  private static List<Type> minimalCandidates(List<Type> erasedCandidates) {
+    List<Type> results = new ArrayList<>();
+    for (Type v : erasedCandidates) {
+      boolean isValid = true;
+      for (Type w : erasedCandidates) {
+        if (!w.equals(v) && w.isSubtypeOf(v)) {
+          isValid = false;
+          break;
+        }
+      }
+      if (isValid) {
+        results.add(v);
+      }
+    }
+    return results;
+  }
+
+  /**
+   * For any element G of MEC that is a generic type, let the "relevant" parameterizations of G, Relevant(G), be:
+   * Relevant(G) = { V | 1 ≤ i ≤ k: V in ST(Ui) and V = G<...> }
+   *
+   * @param minimalErasedCandidates MEC
+   * @param supertypes
+   * @return the set of known parameterizations for each generic type G of MEC 
+   */
+  private static Multimap<Type, Type> relevantParameterizations(List<Type> minimalErasedCandidates, List<Set<Type>> supertypes) {
+    Multimap<Type, Type> result = HashMultimap.create();
+    for (Set<Type> supertypesSet : supertypes) {
+      for (Type supertype : supertypesSet) {
+        Type erasedSupertype = supertype.erasure();
+        if (minimalErasedCandidates.contains(erasedSupertype)) {
+          result.put(erasedSupertype, supertype);
+        }
+      }
+    }
+    return result;
+  }
+
+  @VisibleForTesting
+  static Type best(List<Type> minimalCandidates) {
+    Collections.sort(minimalCandidates, (t1, t2) -> {
+      // Sort minimal candidates by name with classes before interfaces, to guarantee always the same type is returned when approximated.
+      Symbol.TypeSymbol t1Symbol = t1.symbol();
+      Symbol.TypeSymbol t2Symbol = t2.symbol();
+      if (t1Symbol.isInterface() && t2Symbol.isInterface()) {
+        return t1.name().compareTo(t2.name());
+      } else if (t1Symbol.isInterface()) {
+        return 1;
+      } else if (t2Symbol.isInterface()) {
+        return -1;
+      }
+      return t1.name().compareTo(t2.name());
+    });
+    // FIXME SONARJAVA-1632 should return union of types
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)